## Probe Match

In [1]:
%pwd

'/home/junyuchen/Lab/Probe-Match'

### Prokka - Contigs annotation

In [ ]:
prokka --outdir mydir --prefix mygenome contigs.fa

In [ ]:
prokka --outdir Root11 --prefix Root11 Root11.fna

```shell
for fasta in `ls | grep .fna`
do
    prokka --outdir /home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/${fasta:0:-4} --prefix ${fasta:0:-4} $fasta
done
```

Parising GFF file

#### bcbio-gff
conda install -c bioconda bcbio-gff

In [1]:
from BCBio import GFF

In [2]:
%pwd

'/home/junyuchen/Lab/Probe-Match'

```python
from BCBio import GFF

in_file = "your_file.gff"

limit_info = dict(
        gff_id = ["chr1"],
        gff_source = ["Coding_transcript"])

in_handle = open(in_file)
for rec in GFF.parse(in_handle, limit_info=limit_info):
    print rec.features[0]
in_handle.close()
```

In [ ]:
product=23S ribosomal RNA
product=16S ribosomal RNA

In [122]:
#in_file = "/home/junyuchen/Lab/Probe-Match/Root-WGS/Root11/Root11.gff"
in_file = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2/Root123D2.gff"

limit_info = dict(gff_type = ["rRNA"])
        

in_handle = open(in_file)
for record in GFF.parse(in_handle, limit_info=limit_info):
    #print(type(rec.features))
    #print(rec.features[0].qualifiers["product"])
    #print(rec.id)
    #print(rec.features[0])
    print("ID = %s, length %i, with %i features"
          % ( record.id, len(record.seq), len(record.features)))
    print (record.features)
    
in_handle.close()

ID = contig_1, length 1914680, with 0 features
[]
ID = contig_2, length 1264344, with 0 features
[]
ID = contig_3, length 926414, with 3 features
[SeqFeature(FeatureLocation(ExactPosition(67862), ExactPosition(67971), strand=-1), type='rRNA', id='MOFEJJPE_03177'), SeqFeature(FeatureLocation(ExactPosition(68060), ExactPosition(70869), strand=-1), type='rRNA', id='MOFEJJPE_03178'), SeqFeature(FeatureLocation(ExactPosition(71675), ExactPosition(73160), strand=-1), type='rRNA', id='MOFEJJPE_03181')]
ID = contig_4, length 409661, with 0 features
[]
ID = contig_5, length 209741, with 0 features
[]
ID = contig_6, length 59929, with 0 features
[]
ID = contig_7, length 14163, with 0 features
[]
ID = contig_8, length 5484, with 0 features
[]


In [164]:
in_file = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2/Root123D2.gff"

In [56]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.SeqFeature import SeqFeature, FeatureLocation
in_file = "/home/junyuchen/Lab/Probe-Match/Root-WGS/Root11/Root11.gff"

limit_info = dict(gff_type = ["rRNA"])
        

in_handle = open(in_file)
for record in GFF.parse(in_handle, limit_info=limit_info):
    #print(type(rec.features))
    #print(rec.features[0].qualifiers["product"])
    #print(rec.id)
    #print(rec.features[0])
    #print("ID = %s, length %i, with %i features"
    #      % ( record.id, len(record.seq), len(record.features)))
    #print(record.features)
    #print (len(record.features))
    if len(record.features) != 0:
        for i in range(len(record.features)):
            #print(record.id)
            #print(record.features[0])
            #print(len(record.seq))
            #print(record.features[i-1].qualifiers["product"])
            #print(type(record.features[i-1].qualifiers["product"]))
            if record.features[i-1].qualifiers["product"] == ['23S ribosomal RNA']:
                print(record.features[i-1])
                print(record.id)
                print(record.features[i-1].location.strand)
                print(record.features[i-1].location.start)
                print(record.features[i-1].location.end)
                print(type(record.features[i-1].location))
                #feature_seq = record.seq[record.features[i-1].location.start:record.features[i-1].location.end].reverse_complement()
                #print(feature_seq)
                seq_23S = SeqRecord(record.features[i-1].location.extract(record.seq),\
                                    id=record.features[i-1].id,\
                                    description=str(record.id +'-'+ record.features[i-1].qualifiers["product"][0]))
                
                #print(seq_23S)
                SeqIO.write(seq_23S, "my_example.faa", "fasta")
            elif record.features[i-1].qualifiers["product"] == ['16S ribosomal RNA']:
                print(record.features[i-1])
                print(record.id)
                seq_16S = record.features[i-1].location.extract(record.seq)
                #print(seq_16S)
                
                
            #print
        #print(record.features)
in_handle.close()

type: rRNA
location: [208745:210295](-)
id: NLKIOJMO_05069
qualifiers:
    Key: ID, Value: ['NLKIOJMO_05069']
    Key: locus_tag, Value: ['NLKIOJMO_05069']
    Key: product, Value: ['16S ribosomal RNA']
    Key: score, Value: ['0']
    Key: source, Value: ['barrnap:0.9']

contig_9
type: rRNA
location: [205678:208597](-)
id: NLKIOJMO_05068
qualifiers:
    Key: ID, Value: ['NLKIOJMO_05068']
    Key: locus_tag, Value: ['NLKIOJMO_05068']
    Key: product, Value: ['23S ribosomal RNA']
    Key: score, Value: ['0']
    Key: source, Value: ['barrnap:0.9']

contig_9
-1
205678
208597
<class 'Bio.SeqFeature.FeatureLocation'>


In [48]:
%pwd

'/home/junyuchen/Lab/Probe-Match'

feature_seq = example_parent[example_feature.location.start:example_feature.location.end].reverse_complement()

feature_seq = example_feature.extract(example_parent)

contig_9	barrnap:0.9	rRNA	205679	208597	0	-	.	ID=NLKIOJMO_05068;locus_tag=NLKIOJMO_05068;product=23S ribosomal RNA

In [1]:
import os

dir = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/"

for root,dirs,files in os.walk(dir):
    #print(dirs)
    for tempFile in files:
        if "gff" in tempFile:
            path = os.path.join(root,tempFile)
            print(path[:-4])
            #print(dirs)
            print(root)
            print(tempFile[:-4])
            
            
            


/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root920/Root920
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root920
Root920
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root265/Root265
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root265
Root265
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root630/Root630
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root630
Root630
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root70/Root70
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root70
Root70
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root473/Root473
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root473
Root473
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1298/Root1298
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1298
Root1298
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root495/Root495
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root495
Root495
/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root170/Roo

In [59]:
limit_info = dict(gff_type = ["rRNA"])
        
def FindrRNA(path, file):
    
    in_handle = open(path)

    for record in GFF.parse(in_handle, limit_info=limit_info):

        if len(record.features) != 0:
            for i in range(len(record.features)):

                if record.features[i-1].qualifiers["product"] == ['23S ribosomal RNA']:
                    #print(record.features[i-1])
                    #print(record.id)
                    #print(record.features[i-1].location.strand)
                    #print(record.features[i-1].location.start)
                    #print(record.features[i-1].location.end)
                    #print(type(record.features[i-1].location))
                    #feature_seq = record.seq[record.features[i-1].location.start:record.features[i-1].location.end].reverse_complement()
                    #print(feature_seq)
                    seq_23S = SeqRecord(record.features[i-1].location.extract(record.seq),\
                                    id=record.features[i-1].id,\
                                    description=str(record.id +'-'+ record.features[i-1].qualifiers["product"][0]))
                
                    #print(seq_23S)
                    #SeqIO.write(seq_23S, path[:-4]+"_23S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_23S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA/"+tempFile[:-4]+"_23S_rRNA.fasta", "fasta")
                elif record.features[i-1].qualifiers["product"] == ['16S ribosomal RNA']:
                    #print(record.features[i-1])
                    #print(record.id)
                    seq_16S = SeqRecord(record.features[i-1].location.extract(record.seq),\
                                    id=record.features[i-1].id,\
                                    description=str(record.id +'-'+ record.features[i-1].qualifiers["product"][0]))
                
                    #SeqIO.write(seq_16S, path[:-4]+"_16S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_16S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA/"+tempFile[:-4]+"_16S_rRNA.fasta", "fasta")
                    #print(seq_16S)


                #print
            #print(record.features)
    in_handle.close()

In [60]:
import os
from BCBio import GFF
from Bio.Seq import Seq
from Bio import SeqIO
from Bio.SeqFeature import SeqFeature, FeatureLocation

dir = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/"

for root,dirs,files in os.walk(dir):
    #print(dirs)
    for tempFile in files:
        if "gff" in tempFile:
            path = os.path.join(root,tempFile)
            FindrRNA(path, tempFile)
            #print(path[:-4])
            #print(dirs)
            #print(root)
            #print(tempFile[:-4])
            
            
            


In [11]:
# -*- coding: utf-8 -*-
"""
Created on Sat Nov  2 09:52:16 2019

@author: Junyu
"""

import os
from BCBio import GFF
from Bio.Seq import Seq
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from Bio.SeqFeature import SeqFeature, FeatureLocation

dir = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/"

limit_info = dict(gff_type = ["rRNA"])
        
def FindrRNA(path, tempFile):
    
    in_handle = open(path)

    for record in GFF.parse(in_handle, limit_info=limit_info):

        if len(record.features) != 0:
            for i in range(len(record.features)):

                if '23S ribosomal RNA' in record.features[i].qualifiers["product"][0]:
                    #print(record.features[i])
                    #print(record.id)
                    #print(record.features[i].location.strand)
                    #print(record.features[i].location)
                    #print(record.features[i].location.start)
                    #print(record.features[i].location.end)
                    #print(type(record.features[i].location))
                    #feature_seq = record.seq[record.features[i].location.start:record.features[i].location.end].reverse_complement()
                    #print(feature_seq)
                    '''
                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=record.features[i].id,\
                                    description=str(record.id +'-'+ record.features[i].qualifiers["product"][0]))
                    '''
                    loc23 = int(record.features[i].location.end) - int(record.features[i].location.start)
                    #print(loc)
                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=tempFile[:-4],\
                                    description=str(record.id +'-'+record.features[i].id+'-'+ record.features[i].qualifiers["product"][0]+'-'+str(loc23)+ "bp"))
                    #print(seq_23S)
                    #print(record.features[i].location.start)
                    #SeqIO.write(seq_23S, path[:-4]+"_23S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_23S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA2/"+tempFile[:-4]+"_23S_rRNA.fasta", "fasta")
                elif '16S ribosomal RNA' in record.features[i].qualifiers["product"][0]:
                    #print(record.features[i])
                    #print(record.id)
                    loc16 = int(record.features[i].location.end) - int(record.features[i].location.start)
                    seq_16S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=tempFile[:-4],\
                                    description=str(record.id +'-'+record.features[i].id+'-'+ record.features[i].qualifiers["product"][0]+'-'+str(loc16)+"bp"))
                
                    #SeqIO.write(seq_16S, path[:-4]+"_16S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_16S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA2/"+tempFile[:-4]+"_16S_rRNA.fasta", "fasta")
                    #print(seq_16S)


                #print
            #print(record.features)
    in_handle.close()

                                        
for root,dirs,files in os.walk(dir):
    #print(dirs)
    for tempFile in files:
        if "gff" in tempFile:
            path = os.path.join(root,tempFile)
            FindrRNA(path, tempFile)
            #print(path[:-4])
            #print(dirs)
            #print(root)
            #print(tempFile[:-4])
            
            

In [13]:
# -*- coding: utf-8 -*-
"""
Created on Sat Nov  2 09:52:16 2019

@author: Junyu
"""

import os
from BCBio import GFF
from Bio.Seq import Seq
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from Bio.SeqFeature import SeqFeature, FeatureLocation

dir = "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/"

limit_info = dict(gff_type = ["rRNA"])
        
def FindrRNA(path, tempFile):
    
    in_handle = open(path)

    for record in GFF.parse(in_handle, limit_info=limit_info):

        if len(record.features) != 0:
            for i in range(len(record.features)):

                if '23S ribosomal RNA' in record.features[i].qualifiers["product"][0]:
                    #print(record.features[i])
                    #print(record.id)
                    #print(record.features[i].location.strand)
                    #print(record.features[i].location)
                    #print(record.features[i].location.start)
                    #print(record.features[i].location.end)
                    #print(type(record.features[i].location))
                    #feature_seq = record.seq[record.features[i].location.start:record.features[i].location.end].reverse_complement()
                    #print(feature_seq)
                    '''
                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=record.features[i].id,\
                                    description=str(record.id +'-'+ record.features[i].qualifiers["product"][0]))
                    '''
                    loc23 = int(record.features[i].location.end) - int(record.features[i].location.start)
                    #print(loc)
                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=tempFile[:-4],\
                                    description=str( str(loc23)+ "bp"))
                    #print(seq_23S)
                    #print(record.features[i].location.start)
                    #SeqIO.write(seq_23S, path[:-4]+"_23S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_23S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA3/"+tempFile[:-4]+"_23S_rRNA.fasta", "fasta")
                elif '16S ribosomal RNA' in record.features[i].qualifiers["product"][0]:
                    #print(record.features[i])
                    #print(record.id)
                    loc16 = int(record.features[i].location.end) - int(record.features[i].location.start)
                    seq_16S = SeqRecord(record.features[i].location.extract(record.seq),\
                                    id=tempFile[:-4],\
                                    description=str(str(loc16)+"bp"))
                
                    #SeqIO.write(seq_16S, path[:-4]+"_16S_rRNA.fasta", "fasta")
                    SeqIO.write(seq_16S, "/home/junyuchen/Lab/Probe-Match/Root-rRNA3/"+tempFile[:-4]+"_16S_rRNA.fasta", "fasta")
                    #print(seq_16S)


                #print
            #print(record.features)
    in_handle.close()

                                        
for root,dirs,files in os.walk(dir):
    #print(dirs)
    for tempFile in files:
        if "gff" in tempFile:
            path = os.path.join(root,tempFile)
            FindrRNA(path, tempFile)
            #print(path[:-4])
            #print(dirs)
            #print(root)
            #print(tempFile[:-4])
            
            